src/bom/bitcode.rs

```diff
@@ -568,38 +547,16 @@ fn obj_already_has_bitcode(cwd : &Path, obj_target : &OsString) -> bool {
 
 /// Given the name (path) of a tar file on disk, use objcopy to inject it into
 /// the distinguished ELF section for holding the bitcode (see `ELF_SECTION_NAME`)
-fn inject_bitcode(chan : &mut mpsc::Sender<Option<Event>>,
-                  cwd : &Path,
-                  orig_target : &OsString,
-                  bc_target : &OsString,
-                  tar_name : &String,
-                  object_path : &OsString) -> anyhow::Result<()> {
-    let mut objcopy_args = Vec::new();
-    objcopy_args.push(OsString::from("--add-section"));
-    objcopy_args.push(OsString::from(format!("{}={}", ELF_SECTION_NAME, tar_name)));
-    objcopy_args.push(object_path.into());
-
-    match process::Command::new("objcopy").args(&objcopy_args).stdout(process::Stdio::piped()).stderr(process::Stdio::piped()).current_dir(cwd).spawn() {
-        Err(msg) => {
-            let objcopy_ver = get_objcopy_version_info();
-            return Err(anyhow::Error::new(BitcodeError::ErrorAttachingBitcode(cwd.to_path_buf(), orig_target.clone(), bc_target.clone(), msg, objcopy_ver)));
-        }
-        Ok(child) => {
-            let out = child.wait_with_output()?;
-            if !out.status.success() {
-                let err = Event::BitcodeAttachError(Path::new("objcopy").to_path_buf(),
-                                                    Vec::from(objcopy_args),
-                                                    out.stdout,
-                                                    out.stderr,
-                                                    out.status.code());
-                let _res = chan.send(Some(err))?;
-            } else {
-                    let _res = chan.send(Some(Event::BitcodeCaptured(Path::new(bc_target).to_path_buf())));
-            }
-
-            Ok(())
-        }
-    }
+fn inject_bitcode(bc_args : &mut BitcodeArguments) -> anyhow::Result<()> {
+    let objcopy = bc_args.ops.push_op(
+        SubProcOperation::new(
+            &"objcopy",
+            &FileSpec::Replace(String::from("INPFILE"), NamedFile::temp("")),
+            &FileSpec::Append(NamedFile::temp(".o"))));
+
+    objcopy.push_arg("--add-section");
+    objcopy.push_arg(format!("{}=INPFILE", ELF_SECTION_NAME));
```

Comment:
I had another visceral reaction here. I suppose I should have understood it from looking at the API of the pipeline library, but it is more clear here that this is string replacement.

Can you help me understand the big picture of the API here? It feels like this is trying to separate arguments from formatting, but I don't quite get it. Is there space for a more unified builder-ey API like:

```
PipeOperation::new("objcopy")
  .arg("--add-section")
  .arg_with_input(|inp_file| format!("{}={}", ELF_SECTION_NAME, inp_file))
  .arg(NamedFile::temp(".o"))
  .build()
```

Reply:
This is building information for a command that will be executed as part of a chain later, so some of the things (like actual filenames) won't be known until the chain is executed.  Thus the `.build()`  portion of your example would happen at some later point.

My approach was a bit more declarative in nature, whereas what you outline above is based more around callbacks (which I did not start out using, until I introduced the `calling` declaration) because the initial API was much simpler and subsequent growth continued to follow that more declarative approach.  I can rework the API along the example above with more callbacks.  I'll look at more of a callback-based approach along with potentially changing the `calling` along the same lines and see what I can come up with.  Give me a day or so to work on this and I'll post updates here.

Reply:
I appreciate that some of the inputs don't exist yet. I was thinking that `.build()` might not be creating a command, but an AST - that AST might have that lambda (unevaluated) in it so that it can be filled in later during execution. On reflection I see that you do need to be able to declare what your output(s) are, especially for cases where there can be multiple that might need to be referenced in different places.

What do you think about having a system like `NamedFile`, but where each "file" has a globally-unique identity so that it can be referred to in multiple piped operations? Something like:

```
// Where the constructor adds a unique id to this identifier (tracked by a global lazy mutable once_map or something)
let objcopy_output = NamedFile::temp(".o");
PipeOperation::new("objcopy")
  .arg(objcopy_output)
  .build();

PipeOperation::new("md5sum")
  .arg(objcopy_output)
  .build();
```

Does that help with some of the concerns that `FileSpec` is addressing?